### PR TITLE
Fixed TaskHubName to work regardless of get/set order

### LINKED ISSUE #1057: Task Hub name must be specified in host.json when using slots
**Describe the bug**

When I deploy a Durable Functions App to Function App, it said

> The function runtime is unable to start. Microsoft.Azure.WebJobs.Extensions.DurableTask: Task Hub name must be specified in host.json when using slots.

**Investigative information**

Durable Functions extension version: 2.0.0
Function App version (1.0 or 2.0): 2.0
Programming language used: C#

My host.json file content:

```
{
  "version": "2.0",
  "extensions": {
    "durableTask": {
      "hubName": "%SlotTaskHub%"
    }
  }
}
```

My settings.json _**(main slot)**_:

```
{
  "IsEncrypted": false,
  "Values": {
    "SlotTaskHub": "MainHubName"
  }
}
```

My settings.json _**(dev slot)**_:

```
{
  "IsEncrypted": false,
  "Values": {
    "SlotTaskHub": "DevHubName"
  }
}
```

My client function:

```
[FunctionName("MyFunctionName")]
public async Task<HttpResponseMessage> MyFunctionAsync([HttpTrigger(AuthorizationLevel.Anonymous, "POST")]HttpRequestMessage request,
    [DurableClient(TaskHub = "%SlotTaskHub%")]IDurableOrchestrationClient starter)
{
    //My code here
}
```

**Expected behavior**
Successfully works in both slots with different TaskHub names and the same storage account.

**Actual behavior**
If deployed to the main slot, it works.
If deployed to other slots, I receive the error.

**Additional context**

- Development environment (ex. Visual Studio) Visual Studio 2019
- Additional bindings used HttpTrigger or ServiceBusTrigger

### LINKED ISSUE #1068: 2.0.0 Not honouring host.json hubName
### Description

I have a single C# Durable Function app with a staging slot. I have set the hubName in host.json but this does not seem to be honoured.

### Expected behavior
Allow me to run a staging slot against the same storage account.

### Actual behavior
`The function runtime is unable to start. Microsoft.Azure.WebJobs.Extensions.DurableTask: Task Hub name must be specified in host.json when using slots. See documentation on Task Hubs for information on how to set this: https://docs.microsoft.com/azure/azure-functions/durable/durable-functions-task-hubs.
Session Id: 3cf1a9c74a9d465995114003685c9a2a`

I have found DurableTaskOptions.cs which throws `if (IsInNonProductionSlot() && this.isDefaultHubName)`

### Relevant source code snippets

```json
{
    "version": "2.0",
    "extensions": {
        "durableTask": {
            "hubName": "%Version%",
            "azureStorageConnectionStringName": "AzureWebJobsStorage" 
        }
    }
}
```

Version is my build number that I would like to reference from appsettings. 

### Known workarounds
Revert back to V1.8.3

### App Details
- Durable Functions extension version 2.0.0
- Azure Functions runtime version 3.0 (v3-preview)
- Programming language used C#

I also had the same issues when reverting back to netcore2.1 and function v2 so this does not seem to be an issue with v3 preview.

### If deployed to Azure
App name - eoipreprodfunc-westeurope
Func name GetHealthCheck
When 23Nov 9:00 - 14:00 GMT
Storage acc eoipreprodfuncstorage

### PR DESCRIPTION
**Pull request by amdeel**
Changed TaskHubName in DurableTaskOptions to compare to default value instead of using get and set logic to set a bool field.

This fixes many issues with TaskHub names using slots.

resolves #1068 
resolves #1057

### CODE REVIEW

File: src/WebJobs.Extensions.DurableTask/Options/DurableTaskOptions.cs [hunk 1]
```diff
@@ -206,8 +211,8 @@ internal void Validate()
 
             if (IsInNonProductionSlot() && this.IsDefaultHubName())
             {
-                throw new InvalidOperationException("Task Hub name must be specified in host.json when using slots. See documentation on Task Hubs for " +
-                    "information on how to set this: https://docs.microsoft.com/azure/azure-functions/durable/durable-functions-task-hubs");
+                throw new InvalidOperationException($"Task Hub name must be specified in host.json when using slots. Specified name must not equal the defaultHubName ({this.defaultHubName})." +
```
Comment (cgillum) [inline src/WebJobs.Extensions.DurableTask/Options/DurableTaskOptions.cs]: I think people will be confused when they see `defaultHubName` in this message because that's the name of a some private variable that they won't know anything about. I suggest rewording this without referring to the word "defaultHubName".

File: src/WebJobs.Extensions.DurableTask/Options/DurableTaskOptions.cs [hunk 2]
```diff
@@ -225,12 +230,7 @@ internal void Validate()
 
         internal bool IsDefaultHubName()
         {
-            return this.isDefaultHubName ?? this.hubName == null;
-        }
-
-        internal void SetDefaultHubName(string defaultHubName)
-        {
-            this.hubName = defaultHubName;
+            return string.Equals(this.defaultHubName, this.hubName);
```
Comment (cgillum) [inline src/WebJobs.Extensions.DurableTask/Options/DurableTaskOptions.cs]: Task hub names are case insensitive, so you should use a case-insensitive string comparison here.

File: src/WebJobs.Extensions.DurableTask/Options/DurableTaskOptions.cs [hunk 3]
```diff
@@ -152,6 +147,16 @@ public string HubName
         // Used for mocking the lifecycle notification helper.
         internal HttpMessageHandler NotificationHandler { get; set; }
 
+        /// <summary>
+        /// Sets HubName to a value that is considered a default value.
+        /// </summary>
+        /// <param name="hubName">TaskHub name that is considered the default.</param>
+        public void SetDefaultHubName(string hubName)
```
Comment (cgillum) [inline src/WebJobs.Extensions.DurableTask/Options/DurableTaskOptions.cs]: Does this need to be `public`, or can it be `internal`?

Comment (amdeel) [reply src/WebJobs.Extensions.DurableTask/Options/DurableTaskOptions.cs]: I listed it as public so that any other DurabilityProviders can set the default as necessary, as we are currently doing in the AzureStorageDurabilityProviderFactory.

**Comment by mpaul31** [comment]
do you guys think this will get merged here soon? right now we are on hold because we are unable to use stage slots until this gets fixed.

**Comment by ConnorMcMahon** [comment]
@mpaul31, this should get merged early this week, and go out with 2.1, which will be release Friday or early next week.

**Comment by ajk2018** [comment]
Is 2.1, with this fix included,  still planned to go out this week?

**Comment by cgillum** [comment]
@ajk2018 Yes. Just one more PR to be merged and then we will do the release.
